File: frontend/src/lib/gatilhos.ts
import { backendUrl } from "./backend";

export type CombinadorCondicoes = "e" | "ou";
export type MetricaGatilho = "asr_percentual" | "acd_segundos" | "pdd_medio_segundos";
export type PeriodoReferenciaGatilho = "media_ontem" | "media_semanal" | "media_mensal";
export type DirecaoGatilho = "aumento" | "queda" | "qualquer";

export type CondicaoGatilho = {
  id: number;
  metrica: MetricaGatilho;
  periodo_referencia: PeriodoReferenciaGatilho;
  direcao: DirecaoGatilho;
  percentual_limite: number;
};

export type Gatilho = {
  id: number;
  nome: string;
  cliente_id: number | null;
  combinador: CombinadorCondicoes;
  ativo: boolean;
  criado_em: string;
  atualizado_em: string;
  condicoes: CondicaoGatilho[];
};

export type GatilhosResponse = {
  registros: number;
  gatilhos: Gatilho[];
};

/**
 * Lê do backend/ as regras de gatilho — globais sempre incluídas; passando `clienteId` soma as
 * individuais daquele cliente. Chamado sempre do servidor (Server Component).
 */
export async function getGatilhos(clienteId?: number): Promise<GatilhosResponse> {
  const params = new URLSearchParams();
  if (clienteId !== undefined) params.set("cliente_id", String(clienteId));

  const res = await fetch(backendUrl(`/api/gatilhos?${params.toString()}`), { cache: "no-store" });
  if (!res.ok) {
    throw new Error(`Backend respondeu ${res.status} ao buscar /api/gatilhos`);
  }
  return (await res.json()) as GatilhosResponse;
}

File: backend/app/schemas/gatilhos.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.db.models import CombinadorCondicoes, DirecaoGatilho, MetricaGatilho, PeriodoReferenciaGatilho


class CondicaoGatilhoCreate(BaseModel):
    metrica: MetricaGatilho
    periodo_referencia: PeriodoReferenciaGatilho
    direcao: DirecaoGatilho
    percentual_limite: float = Field(..., gt=0, description="Limite de variação percentual, ex. 20.0 = 20%")


class CondicaoGatilhoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    metrica: MetricaGatilho
    periodo_referencia: PeriodoReferenciaGatilho
    direcao: DirecaoGatilho
    percentual_limite: float


class GatilhoCreate(BaseModel):
    nome: str
    cliente_id: int | None = Field(
        None, description="Nulo = gatilho global (vale pra todo cliente); preenchido = individual daquele cliente"
    )
    combinador: CombinadorCondicoes = CombinadorCondicoes.E
    ativo: bool = True
    condicoes: list[CondicaoGatilhoCreate] = Field(..., min_length=1)


class GatilhoUpdate(BaseModel):
    nome: str
    combinador: CombinadorCondicoes
    ativo: bool
    condicoes: list[CondicaoGatilhoCreate] = Field(..., min_length=1)


class GatilhoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nome: str
    cliente_id: int | None = None
    combinador: CombinadorCondicoes
    ativo: bool
    criado_em: datetime
    atualizado_em: datetime
    condicoes: list[CondicaoGatilhoResponse]


class GatilhosResponse(BaseModel):
    registros: int
    gatilhos: list[GatilhoResponse]
